src/ai-jsx/util.ts

```diff
@@ -0,0 +1,372 @@
+    }
+
+    messages.push(...makeMemoizedMessagesArray(child, components));
+  }
+
+  return messages;
+}
+
+/**
+ * Keep the last message from each memoizedId.
+ */
+function makeMessagesArrayFrom(
+  span: ABSpan,
+  components: AnyComponent[],
+): AIMessage[] {
+  const memoized = makeMemoizedMessagesArray(span, components);
+
+  const seenMemoizedIds = new Set<string>();
+  const dedupedMessages: AIMessage[] = [];
+
+  for (const memo of memoized.reverse()) {
+    if (memo.memoizedId === undefined) {
+      dedupedMessages.push(memo.message);
+      continue;
+    }
+
+    if (seenMemoizedIds.has(memo.memoizedId)) {
+      continue;
+    }
+
+    seenMemoizedIds.add(memo.memoizedId);
+    dedupedMessages.push(memo.message);
+  }
+
+  return dedupedMessages.reverse();
+}
+
+function makeTemplatesForCompletion(
+  trackerId: string,
+  completionSpan: ABSpan,
+): PromptTracking | undefined {
+  const tracking: PromptTracking = {
+    id: trackerId,
+    templates: [],
+  };
+
+  const seenMemoizedIds = new Set<string>();
+
+  const walk = (span: ABSpan) => {
+    if (
+      span.children.length > 0 &&
+      (span.name === makeComponentName(SystemMessage) ||
+        span.name === makeComponentName(UserMessage))
+    ) {
+      if (span.memoizedId !== undefined) {
+        if (seenMemoizedIds.has(span.memoizedId)) {
+          return;
+        }
+        seenMemoizedIds.add(span.memoizedId);
+      }
+
+      const suffix =
+        span.name === makeComponentName(SystemMessage) ? 'system' : 'user';
+
+      tracking.templates.push({
+        id: `${trackerId}/${suffix}`,
+        template: makeTemplateString(span),
+      });
+    }
+
+    for (const child of span.children) {
+      if (typeof child === 'string') {
+        continue;
+      }
+
+      // Stop if we encounter another chat component
+      if (child.name === makeComponentName(OpenAIChatModel)) {
+        continue;
+      }
+
+      walk(child);
+    }
+  };
+
+  walk(completionSpan);
+
+  return tracking;
+}
+
+export async function processCompletedRootSpan(rootSpan: ABSpan) {
+  const traceId = rootSpan.id;
+  const events: { message: string; args: SendEventArgs }[] = [];
+  const seenTrackerIds = new Set<string>();
+
+  const walk = (span: ABSpan, parentCompletionSpanId?: string) => {
+    let completionSpanId: string | undefined = parentCompletionSpanId;
+
+    if (span.name === makeComponentName(OpenAIChatModel)) {
+      completionSpanId = span.id;
+      const trackerId = parseTrackerIdFromProps(span.props);
+
+      if (!trackerId || !seenTrackerIds.has(trackerId)) {
+        // TODO: Need to figure out how to differentiate between which messages were
+        // sent in the request and which were part of the response, since they're all
+        // rendered together as children under <ChatCompletion>. For now just assuming
+        // that all system and user messages are part of the request and all assistant
+        // messages are part of the response.
```

Comment:
I think it is safe to assume that the last assistant message will always be the response. I don't think you can send in a request with an assistant as the last message.

Reply:
i think assistant messages can be in the request if you're having a conversation

https://platform.openai.com/docs/api-reference/chat/create

<img width="642" alt="Screenshot 2023-11-27 at 11 57 57 AM" src="https://github.com/autoblocksai/javascript-sdk/assets/7498009/b207a4ed-ab66-417e-ac53-69cac4d30baf">

Reply:
Yes they can be in the request, but I've noticed in playground they can't be the last message

Reply:
i see, ok that makes sense